### PR TITLE
Add App Insights via CDK

### PR DESCRIPTION
**Pull request by JoshLove-msft**
###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2929)

### CODE REVIEW

File: src/Aspire.Hosting.Azure/Extensions/AzureApplicationInsightsExtensions.cs [hunk 1]
```diff
@@ -19,10 +24,45 @@ public static class AzureApplicationInsightsExtensions
...
+
+    /// <summary>
+    /// Adds an Azure Application Insights resource to the application model.
+    /// </summary>
+    /// <param name="builder">The builder for the distributed application.</param>
+    /// <param name="name">The name of the resource.</param>
+    /// <param name="configureResource">Optional callback to configure the Application Insights resource.</param>
+    /// <returns></returns>
+    [RequiresPreviewFeatures]
+    public static IResourceBuilder<AzureApplicationInsightsResource> AddAzureApplicationInsights(this IDistributedApplicationBuilder builder, string name, Action<IResourceBuilder<AzureApplicationInsightsResource>, ResourceModuleConstruct>? configureResource = null)
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Azure/Extensions/AzureApplicationInsightsExtensions.cs:40]: This callback should expose the application insights resource and the log analytics workspace resources from the CDK on the callback. The ability to easily get at the resources we've injected to the construct is what makes these callbacks useful.

File: src/Aspire.Hosting.Azure/Extensions/AzureApplicationInsightsExtensions.cs [hunk 2]
```diff
@@ -19,10 +24,45 @@ public static class AzureApplicationInsightsExtensions
...
+    {
+        var configureConstruct = (ResourceModuleConstruct construct) =>
+        {
+            var workspaceIdParameter = new Parameter("logAnalyticsWorkspaceId", defaultValue: "");
+            construct.AddParameter(workspaceIdParameter);
+
+            var opInsights = new OperationalInsightsWorkspace(construct, name: name, sku: new OperationalInsightsWorkspaceSku(OperationalInsightsWorkspaceSkuName.PerGB2018));
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Azure/Extensions/AzureApplicationInsightsExtensions.cs:47]: Creation of this resource should be conditional on whether the `logAnalyticsWorkspaceId` is provided?

File: src/Aspire.Hosting.Azure/Extensions/AzureApplicationInsightsExtensions.cs [hunk 3]
```diff
@@ -19,10 +24,45 @@ public static class AzureApplicationInsightsExtensions
...
+
+            appInsights.AddOutput("appInsightsConnectionString", p => p.ConnectionString);
+        };
+        var resource = new AzureApplicationInsightsResource(name, configureConstruct);
+
         return builder.AddResource(resource)
-                .WithParameter("appInsightsName", resource.CreateBicepResourceName())
-                .WithParameter(AzureBicepResource.KnownParameters.LogAnalyticsWorkspaceId)
-                .WithManifestPublishingCallback(resource.WriteToManifest);
+                      // These ambient parameters are only available in development time.
```
Comment (mitchdenny) [inline src/Aspire.Hosting.Azure/Extensions/AzureApplicationInsightsExtensions.cs:63]: Delete this comment.

File: tests/Aspire.Hosting.Tests/Azure/AzureBicepResourceTests.cs [hunk 4]
```diff
@@ -494,7 +553,7 @@ public async Task PublishAsRedisPublishesRedisAsAzureRedisConstruct()
                 value: '${redisCache_p9fE6TK3F.properties.hostName},ssl=true,password=${redisCache_p9fE6TK3F.listKeys(redisCache_p9fE6TK3F.apiVersion).primaryKey}'
               }
             }
-            
+
```
Comment (mitchdenny) [inline tests/Aspire.Hosting.Tests/Azure/AzureBicepResourceTests.cs:556]: What are these whitespace changes about? This concerns me it means your generation was different to what was there previously. Did CDK change some whitespace?

Comment (JoshLove-msft) [reply tests/Aspire.Hosting.Tests/Azure/AzureBicepResourceTests.cs:556]: These are in the tests and not part of the payload. I think it was just my IDE settings.

File: src/Aspire.Hosting.Azure/Extensions/AzureApplicationInsightsExtensions.cs (hunk 1 above)
Comment (JoshLove-msft) [reply src/Aspire.Hosting.Azure/Extensions/AzureApplicationInsightsExtensions.cs:40]: Woops this was a miss.

**Comment by mitchdenny** [comment]
@JoshLove-msft needs conflicts resolved but this looks good!

**Review by mitchdenny — APPROVED** [review]

**Pull request merged**